framework/src/media/demux/mpeg2ts/TSParser.cpp

```diff
@@ -0,0 +1,654 @@
+}
+
+size_t TSParser::sizeOfSpace(void)
+{
+	return mBufferWriter->sizeOfSpace();
+}
+
+size_t TSParser::pushData(uint8_t *buf, size_t size)
+{
+	size_t written = 0;
+	if (mBufferWriter) {
+		written += mBufferWriter->write(buf, size);
+		assert(written == size);
+	}
+	return written;
+}
+
+size_t TSParser::pullData(uint8_t *buf, size_t size, prog_num_t progNum)
+{
+	if (mPESPid == -1) {
+		// setup PES pid
+		if (progNum == 0) {
+			// use default 1st program
+			std::vector<prog_num_t> programs;
+			mParserManager->GetPrograms(programs);
+			progNum = programs[0];
+		}
+		uint8_t streamType;
+		if (!mParserManager->GetAudioStreamInfo(progNum, streamType, mPESPid)) {
+			meddbg("get audio pes pid failed\n");
+			return 0;
+		}
+
+		medvdbg("setup audio pes pid: 0x%x\n", mPESPid);
+	}
+
+	size_t fill = 0;
+	size_t need;
+	while (fill < size) {
+		need = size - fill;
+		if (mPESPacket) {
+			// get remaining payload in last PES packet
+			if (need > mPESParser->ESDataLength() - mPESDataUsed) {
+				need = mPESParser->ESDataLength() - mPESDataUsed;
+			}
+
+			memcpy(&buf[fill], mPESParser->ESData() + mPESDataUsed, need);
+			mPESDataUsed += need;
+			fill += need;
+
+			if (mPESDataUsed == mPESParser->ESDataLength()) {
+				// read whole es data in one pes packet
+				delete mPESPacket;
+				mPESPacket = NULL;
+				mPESDataUsed = 0;
+			}
+			continue;
+		}
+
+		// get new PES packet
+		if (getPESPacket(&mPESPacket) <= 0) {
+			// 0 / -1
+			break;
+		}
+
+		// parse PES packet
+		if (mPESParser->Parse(mPESPacket->Data(), mPESPacket->DataLength())) {
+			mPESDataUsed = 0;
+		} else {
+			meddbg("PES parse failed\n");
+			assert(0);
+		}
+	}
+
+	return fill;
+}
+
+bool TSParser::getPrograms(std::vector<prog_num_t> &progs)
+{
+	return mParserManager->GetPrograms(progs);
+}
+
+audio_type_t TSParser::getAudioType(prog_num_t progNum)
+{
+	uint8_t streamType;
+	ts_pid_t streamPid;
+	if (mParserManager->GetAudioStreamInfo(progNum, streamType, streamPid)) {
+		switch (streamType) {
+		case PMTElementary::STREAM_TYPE_AUDIO_AAC:
+		case PMTElementary::STREAM_TYPE_AUDIO_HE_AAC:
+			return AUDIO_TYPE_AAC;
+		case PMTElementary::STREAM_TYPE_AUDIO_MPEG1:
+			return AUDIO_TYPE_MP3;
+		default:
+			meddbg("unsupported audio type 0x%x\n", streamType);
+			break;
+		}
+	} else {
+		meddbg("didn't get any audio stream info\n");
+	}
+
+	return AUDIO_TYPE_INVALID;
+}
+
+int TSParser::Adjust(uint8_t *pPacketData, size_t readOffset)
+{
+	uint8_t buffer[TS_PACKET_SIZE];
+	TSHeader tsHeader;
+	size_t szRead;
+	int syncOffset;
+	int count;
+
+	for (syncOffset = 0; syncOffset < TS_PACKET_SIZE; syncOffset++) {
+		if (pPacketData[syncOffset] != SYNCCODE) {
+			continue;
+		}
+
+		// found sync byte, now do sync verification!
+		for (count = 1; count < SYNC_COUNT; count++) {
+			szRead = mBufferReader->copy(buffer, TS_PACKET_SIZE, readOffset + syncOffset + count * TS_PACKET_SIZE);
+			if (szRead != TS_PACKET_SIZE) {
+				// data in buffer is not enough for sync verification
```

Comment:
Adjust is used to sync forcely, then it should wait until it ready to copy instead of return error, at lease try count is needed, I think.
Could you please answer?

Reply:
I use `mBufferReader->copy()` here, `copy` method would not be blocked.
If there's not more data for sync, return error code: DEMUXER_ERROR_WANT_DATA
PS: `Adjust` has been renamed to `resync`